File: ReactNativeApp/android/app/src/main/java/com/components/detection/person/bodyPartDetection/BodyPartState.java
package com.components.detection.person.bodyPartDetection;

import androidx.collection.ArraySet;

import com.components.detection.ObjectDetectionState;
import com.components.detection.person.bone.BoneState;
import com.logger.SLOG;
import com.models.ModelManager;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public abstract class BodyPartState extends ObjectDetectionState {
    // An Bodypart, consisting of multiple bodyElements
    public final List<ObjectDetectionState> bodyElements = new ArrayList<>();
    public final List<BoneState> bones = new ArrayList<>();
    protected Orientation orientation;
    Set<String> associatedLabels = null;
    private Set<ObjectDetectionState> objectDetectionSet;


    public BodyPartState(String label, ModelManager.MODELTYPE modelType, boolean exerciseLead, Orientation orientation) {
        super(label, modelType, exerciseLead);
        this.orientation = orientation;
    }

    public BodyPartState(String label, ModelManager.MODELTYPE modelType, boolean exerciseLead) {
        this(label, modelType, exerciseLead, Orientation.LEFT);
    }

    public List<ObjectDetectionState> getBodyElements() {
        return bodyElements;
    }


    @Override
    public void setConfidenceScore(float confidenceScore) {
        super.setConfidenceScore(confidenceScore);
        bodyElements.forEach(bodyElement -> bodyElement.setConfidenceScore(confidenceScore));
    }

    public Orientation getOrientation() {
        return orientation;
    }

    public Set<String> getAssociatedLabels() {
        //ensures we do it only once
        if (associatedLabels == null) {
            associatedLabels = new ArraySet<>();

            bodyElements.forEach(bodyElement -> associatedLabels.add(bodyElement.getLabel()));
            SLOG.d("labels of " + getName() + " : " + associatedLabels);
        }
        return associatedLabels;
    }

    public Set<ObjectDetectionState> getObjectDetections() {
        if (objectDetectionSet == null) {
            objectDetectionSet = new ArraySet<>();
            objectDetectionSet.addAll(bodyElements);
        }
        //ensures we do it only once
        return objectDetectionSet;
    }

    public abstract String getName();


    public void writeToJSON(JSONObject dataJSONWriter, int locIdx) {
        bodyElements.forEach(element -> {
            try {
                element.writeToJSON(dataJSONWriter);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        });
    }

    @Override
    public void unSubscribe() {
        super.unSubscribe();
        bodyElements.forEach(ObjectDetectionState::unSubscribe);
    }

    public enum Orientation {
        LEFT, RIGHT
    }
}

File: ReactNativeApp/ios/Modules/SSVC/Instance/Detection/Person/BodyPart/BodyPartState.swift
//
//  BodyPartState.swift
//  jogo
//
//  Created by Mohsin on 18/10/2021.
//

import Foundation

public class BodyPartState: ObjectDetectionState {
  
  var bodyElements: [ObjectDetectionState] = []
  var bones: [BoneState] = []
  var associatedLabels: Set<String>?
  var objectDetectionSet: Set<ObjectDetectionState>?
  
  init(label: String, modelType: ModelManager.MODELTYPE, exerciseLead: Bool, observerId: Int) {
    super.init(label: label, modelType: modelType, exerciseLead: exerciseLead, observerId: observerId)
  }
  
  public func getBodyElements() -> [ObjectDetectionState] {
    return bodyElements
  }
  
  public override func setConfidenceScore(confidenceScore: Float) {
    super.setConfidenceScore(confidenceScore: confidenceScore)
    bodyElements.forEach({$0.setConfidenceScore(confidenceScore: confidenceScore)})
  }
  
  public func getAssociatedLabels() -> Set<String> {
    if associatedLabels == nil {
      associatedLabels = Set<String>()
      bodyElements.forEach { (bdyElement) in
        associatedLabels?.insert(bdyElement.label)
      }
    }
    return associatedLabels!
  }
  
  public func getObjectDetections() -> Set<ObjectDetectionState> {
    if objectDetectionSet == nil {
      objectDetectionSet = Set<ObjectDetectionState>()
      bodyElements.forEach({objectDetectionSet?.insert($0)})
    }
    return objectDetectionSet!
  }
  
  public func getName() -> String {
    return ""
  }
  
  public override func unsubscribe() {
    super.unsubscribe()
    bodyElements.forEach({$0.unsubscribe()})
  }
  
  public override func cleanup() {
    super.cleanup()
    bodyElements.forEach({$0.cleanup()})
    self.bodyElements = []
    self.bones = []
    self.associatedLabels = nil
    self.objectDetectionSet = nil
  }
}
